cmd/tracee-ebpf/main.go

```diff
@@ -206,15 +207,35 @@ func main() {
 					}
 					bpfFilePath = "embedded-core"
 					bpfBytes, err = unpackCOREBinary()
-				} else { // build non CO-RE BPF
-					if debug {
-						fmt.Println("BPF: no BTF file was found or provided, building BPF object")
-					}
-					if bpfFilePath, err = getBPFObjectPath(); err != nil {
-						return err
+					if err != nil {
+						return fmt.Errorf("could not unpack embedded CO-RE eBPF object: %v", err)
 					}
-					if bpfBytes, err = ioutil.ReadFile(bpfFilePath); err != nil {
-						return err
+				} else {
+					// check embedded BTFHub to use CO-RE
+					unpackBTFFile := filepath.Join(traceeInstallPath, "/tracee.btf")
+					err = unpackBTFHub(unpackBTFFile, OSInfo)
+					if err != nil {
+						// no other solution than to build non CO-RE object
+						if debug {
+							fmt.Println("BPF: no BTF file was found or provided, building BPF object")
+						}
+						if bpfFilePath, err = getBPFObjectPath(); err != nil {
+							return err
+						}
+						if bpfBytes, err = ioutil.ReadFile(bpfFilePath); err != nil {
+							return err
+						}
+					} else {
```

Comment:
These lines are about to be changed by another PR I have ready (dropping the getBPFOjectPath() building logic and having bpf-nocore built and installed by Makefile (or docker builder)).

Mind if we do that change in that PR ? I can then have 1 commit cleaning this and the 2nd commit dropping the build logic to the builder.

Reply:
Nevermind, I'll improve that in this PR.. this way its cleaner for the next one.

Reply:
I moved that extensive logic into a function called `dealWithGivenEnv()`, and organized the decision taking logic into 5 ordered steps, enumerating where they are.

The `if debug {}` snippets make this look more complex than it really is, but I think, with steps enumerated, it got easier to understand/read. 

Of course, there is always space for improvement AND it will be easier to simplify this once "getBPFOjectPath()" is removed (my next PRs, almost ready).

Reply:
BTW if you think there is a better way of doing this (with goto's) please could you recommend it as a snippet here ? Or if you prefer to do it later, ok too. Your call.

Reply:
Okay, the overall result is better indeed. By using goto's, like you said, we're making it easier to read and also removing a bunch of indentation (which I prefer as well). Nice tip.